### PR TITLE
CASSANDRA-19031: Fix bulk writing when using identifiers that need quotes

### PR DESCRIPTION
**Pull request by frankgh**
Cassandra treats all identifiers as lower case unless explicitly quoted by the users,
(i.e. keyspace names, table names, column names, etc). We can define a case-sensitive
identifier or we can use a reserved word as an identifier by quoting it during DDL
creation.

In the analytics library, bulk writing fails when we encounter these identifiers. In
this commit, we fix the issue by property propagating the information about whether
identifiers need to be quoted by exposing a new dataframe option (`quote_identifiers`).
When set to `true`, it will _maybe_ quote the keyspace/table/column names and it will
properly be able to write data when using mixed-case or reserved words in the
identifiers.

### CODE REVIEW

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/CassandraClusterInfo.java [hunk 1]
```diff
@@ -535,6 +548,15 @@ protected boolean isReplacement(RingEntry ringEntry)
         return false;
     }
 
+    protected CassandraBridge bridge()
+    {
+        if (bridge == null)
+        {
+            bridge = CassandraBridgeFactory.get(getLowestCassandraVersion());
```
Comment (frankgh) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/CassandraClusterInfo.java]: underlying ConcurrentHashMap in factory guarantees single bridge initialization here, so no need for synchronization

Comment (yifan-c) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/CassandraClusterInfo.java]: In that case, why not just call `CassandraBridgeFactory.get(getLowestCassandraVersion())`? Given the value effectively cached already.

File: cassandra-analytics-core/src/test/java/org/apache/cassandra/spark/bulkwriter/TableSchemaTest.java [hunk 2]
```diff
@@ -231,4 +234,10 @@ private TableSchemaTestCommon.MockTableSchemaBuilder getValidSchemaBuilder()
                 .withWriteMode(WriteMode.INSERT)
                 .withDataFrameSchema(validDataFrameSchema);
     }
+
+    // Removes the unique table name to make validation consistent
+    private static String replaceTableName(String statement)
```
Comment (yifan-c) [inline cassandra-analytics-core/src/test/java/org/apache/cassandra/spark/bulkwriter/TableSchemaTest.java]: nit: how about rename to `trimTestTableId`?

File: cassandra-analytics-integration-tests/src/test/java/org/apache/cassandra/analytics/QuoteIdentifiersWriteTest.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+ *
+ * <p>These tests exercise a full integration test, which includes testing Sidecar behavior when dealing with quoted
+ * identifiers.
+ */
+@ExtendWith(VertxExtension.class)
+class QuoteIdentifiersWriteTest extends SparkIntegrationTestBase
+{
+    static final int ROW_COUNT = 10_000;
+
+    @CassandraIntegrationTest(nodesPerDc = 3, gossip = true)
```
Comment (yifan-c) [inline cassandra-analytics-integration-tests/src/test/java/org/apache/cassandra/analytics/QuoteIdentifiersWriteTest.java]: Does the test quire 3 nodes? If not, let's use 1 node to consume less resources.

**Review by yifan-c — COMMENTED** [review]
lgtm overall. A few nits. Feel free to ignore.

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/CassandraClusterInfo.java (hunk 1 above)
Comment (frankgh) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/CassandraClusterInfo.java]: yeah, having a local cache saves look ups from the map. But lookups are fast anyway, so I think we can remove the local field used as a cache

File: cassandra-analytics-integration-tests/src/test/java/org/apache/cassandra/analytics/QuoteIdentifiersWriteTest.java (hunk 3 above)
Comment (frankgh) [reply cassandra-analytics-integration-tests/src/test/java/org/apache/cassandra/analytics/QuoteIdentifiersWriteTest.java]: good point, we don't need 3 instances. I will lower it to 1

File: cassandra-analytics-core/src/test/java/org/apache/cassandra/spark/bulkwriter/TableSchemaTest.java (hunk 2 above)
Comment (frankgh) [reply cassandra-analytics-core/src/test/java/org/apache/cassandra/spark/bulkwriter/TableSchemaTest.java]: table Id might cause confusion with Cassandra's table ID. I will rename it to `trimUniqueTableName`

**Comment by frankgh** [comment]
> lgtm overall. A few nits. Feel free to ignore.

Thanks for the review. I've addressed your comments.

**Comment by frankgh** [comment]
Squashed and lowered concurrency for JDK8 tests

**Comment by frankgh** [comment]
Closed via https://github.com/apache/cassandra-analytics/commit/c7c3bbca2c7cb415b39689e924fa2357c239f043

**Pull request closed without merging**